### ISSUE TITLE
Add Module::to_device; layers upload CPU weights to the GPU on every forward and the fused Adam step never runs

### ISSUE THREAD

**Issue by vyncint** [vyncint/oxmera#46]
## Summary

Every `nn` layer moves its parameters to the input's device inside `forward`, on every call, and nothing can move them once. On a GPU input that is a host→device copy of every weight per forward pass, a device→host copy of every gradient per backward pass, and the optimizer's fused GPU step is unreachable because the parameters themselves live on the CPU.

## Where (main at 853a224)

```rust
// crates/oxmera-nn/src/linear.rs:55-61
fn forward(&self, input: &Tensor) -> Result<Tensor> {
    crate::check_param_dtype("Linear", input, &self.parameters())?;
    let w = self.weight.value().to_device(input.device())?;
    let mut y = input.matmul(&w.t()?)?;
    if let Some(bias) = &self.bias {
        let b = bias.value().to_device(input.device())?;
```

Same pattern in `conv.rs:98,113` and `norm.rs:54-55,140-141`. `Module` (`crates/oxmera-nn/src/lib.rs:73-97`) has `forward`, `parameters`, `named_parameters`, `zero_grad`, `set_training` and nothing that moves parameters. The optimizer takes the fused path only when `value.device() != Device::Cpu` (`crates/oxmera-optim/src/lib.rs:188`).

## Reproduction

```rust
let m = Sequential::new().push(Linear::new(2, 2, 1));
let x = Tensor::randn([4, 2]).to_device(Device::Metal { index: 0 })?;
m.forward(&x)?;
assert_eq!(m.parameters()[0].value().device(), Device::Cpu); // holds today
```

`oxmera train --device metal|cuda` (`crates/oxmera-cli/src/train.rs:170`) trains this way. oxmega works around it by building its own modules and moving each `Param` by hand.

## Why it matters

This is the per-step cost `docs/LIMITATIONS.md` describes when it says the GPU is still not faster than the CPU on the motivating benchmark. Any model on Metal or CUDA pays it silently, and the 0.3.0 fused `adam_step` kernel has no caller from `nn`.

## Fix

- `Module::to_device(&self, device: Device) -> Result<()>` with a default implementation that `Param::set`s each parameter to `value.to_device(device)?` (keeping leaf semantics, which `Param::set` already has); `Sequential` forwards to its children; `BatchNorm2d` moves its running statistics too.
- The CLI and `train_mnist` call it once after construction.
- Keep the per-forward `to_device` as the safety net it already is (a no-op clone when devices match).

## Done when

- A test asserts that after `to_device(gpu)` the second `forward` performs no `upload` (count through the backend's transfer counters or a mock backend).
- A test asserts the optimizer takes the fused path for a model moved to a GPU device (runs under `--features hardware` for CUDA; on macOS for Metal).
- LIMITATIONS' Metal paragraph is re-measured after the change and updated with the new numbers, whatever they are.
